components/camel-elasticsearch5-rest/src/main/java/org/apache/camel/component/elasticsearch5/ElasticsearchProducer.java

```diff
@@ -0,0 +1,295 @@
+
+    private static final Logger LOG = LoggerFactory.getLogger(ElasticsearchProducer.class);
+
+    protected final ElasticsearchConfiguration configuration;
+    private RestClient client;
+    private Sniffer sniffer;
+
+    public ElasticsearchProducer(ElasticsearchEndpoint endpoint, ElasticsearchConfiguration configuration) {
+        super(endpoint);
+        this.configuration = configuration;
+        this.client = endpoint.getClient();
+    }
+
+    private ElasticsearchOperation resolveOperation(Exchange exchange) {
+        // 1. Operation can be driven by either (in order of preference):
+        // a. If the body is an ActionRequest the operation is set by the type
+        // of request.
+        // b. If the body is not an ActionRequest, the operation is set by the
+        // header if it exists.
+        // c. If neither the operation can not be derived from the body or
+        // header, the configuration is used.
+        // In the event we can't discover the operation from a, b or c we throw
+        // an error.
+        Object request = exchange.getIn().getBody();
+        if (request instanceof IndexRequest) {
+            return ElasticsearchOperation.INDEX;
+        } else if (request instanceof GetRequest) {
+            return ElasticsearchOperation.GET_BY_ID;
+        } else if (request instanceof MultiGetRequest) {
+            return ElasticsearchOperation.MULTIGET;
+        } else if (request instanceof UpdateRequest) {
+            return ElasticsearchOperation.UPDATE;
+        } else if (request instanceof BulkRequest) {
+            // do we want bulk or bulk_index?
+            if (configuration.getOperation() == ElasticsearchOperation.BULK_INDEX) {
+                return configuration.getOperation().BULK_INDEX;
+            } else {
+                return configuration.getOperation().BULK;
+            }
+        } else if (request instanceof DeleteRequest) {
+            return ElasticsearchOperation.DELETE;
+        } else if (request instanceof SearchRequest) {
+            return ElasticsearchOperation.SEARCH;
+        } else if (request instanceof DeleteIndexRequest) {
+            return ElasticsearchOperation.DELETE_INDEX;
+        }
+
+        ElasticsearchOperation operationConfig = exchange.getIn().getHeader(ElasticsearchConstants.PARAM_OPERATION, ElasticsearchOperation.class);
+        if (operationConfig == null) {
+            operationConfig = configuration.getOperation();
+        }
+        if (operationConfig == null) {
+            throw new IllegalArgumentException(ElasticsearchConstants.PARAM_OPERATION + " value '" + operationConfig + "' is not supported");
+        }
+        return operationConfig;
+    }
+
+    public void process(Exchange exchange) throws Exception {
+        if (configuration.getDisconnect() && client == null) {
+            startClient();
+        }
+        RestHighLevelClient restHighLevelClient = new RestHighLevelClient(client);
```

Comment:
I wonder if we should create a new `RestHighLevelClient` per message? Or is this just a facade and okay to do.

Reply:
High level REST client wraps an instance of the low level RestClient and allows to build requests and read responses. 
So I think it's okay to create an instance per message